### DISCUSSION TITLE
Duplicate Structured Data (being picked up by Google from hydration data in source code) showing in Google Rich results Testing Tool

### DISCUSSION THREAD

**Discussion by valley-boy** [vercel/next.js#66896] [Help]
### Summary

Hi, We are seeing Duplicate Structured Data in Google Rich results Testing Tool, pretty sure it is being picked up from the hydration data at the bottom of the source code, 

![image](https://github.com/vercel/next.js/assets/143649490/71b6d15a-4883-4b65-b348-634f05228901)

![image](https://github.com/vercel/next.js/assets/143649490/a1f404ec-83e0-4321-bc3f-1dea7f34a2b8)

e.g. has 1 review, but is showing twice in the testing tool as below image
https://www.nursinghomes.com/ne/omaha/good-samaritan-society-millard/
![image](https://github.com/vercel/next.js/assets/143649490/5763ab0c-b546-4943-8d8d-98e6230dae5f)

This is a significant issue as everything we read says that the testing tool is a fair representation of what Google sees, and the fact that the rich results are not showing in Google search results is testament to this being an issue.

To note we have chosen to put the structured data json-ld in the body vs head due to Next making it arduous to add json-ld to the head

Also to note on the schema.org validation tool, the structured data is not showing as duplicated

![image](https://github.com/vercel/next.js/assets/143649490/58036760-2abc-4968-9870-633e69bca939)

Any ideas much appreciated, thank you

**Comment by leopoldkristjansson** [comment]
Just to chip in, this seems to be an issue in [Nuxt](https://github.com/nuxt) as well. I have some SSR Nuxt v3 projects running on Vercel and I am seeing the same. This probably has nothing to with Vercel though.

This thread is closed and a bit old, but the discussion is relevant: https://github.com/ymmooot/nuxt-jsonld/issues/423

In Nuxt, I am assuming it has something to do with the crawler being able to run JS and hydration being responsible for the issue.

**Reply by leopoldkristjansson** [reply]
I can confirm I solved this for Nuxt using Nuxt's `callOnce` util function. [More on that here](https://github.com/ymmooot/nuxt-jsonld/issues/423#issuecomment-2473200562).

@blf885 maybe/probably Next has a way to achieve the [same functionality as callOnce for Nuxt](https://nuxt.com/docs/api/utils/call-once#purpose)?

**Comment by blf885** [comment]
Has anyone else experienced this? I'm looking at Next JS because it is supposed to help with SEO. A big part of that we will want is good structured data for product pages, but this seems to be an issue with hydration and few people have really said anything on here. Are you still experiencing these issues or have they been solved?

**Reply by m4w** [reply]
Yes, we have exactly the same issue right now. 
We are using the nexjs version 15.0.2 at the moment

**Reply by blf885** [reply]
I'm curious,  I'm still working to understand how the client and server components work,  but if I understand correctly,  you should be able to use a server component to avoid this.  Because the duplicate is coming from the hydration payload, which isn't sent for server components.  Right? So if you have your schema data server side and can use a server component, you should be able to avoid this issue.

**Reply by m4w** [reply]
We are using server components, but there is still this nextjs script which has the structured data included and which gets picked up additionally by google or schema.org/validator.

This is the nextjs script ( a part of it):

`self.__next_f.push([1,"36:[\"$\",\"script\",null,{\"type\":\"application/ld+json\",\"children\":\"{\\\"@type\\\":\\\"Organization\\\",...`

**Reply by m4w** [reply]
But we have a mixed React-Component-Tree which has Client and Server Components nested into each other, which is supported by nextjs.

I guess in the end this results in those scripts.

**Reply by blf885** [reply]
I see. I did some more reading and I understand the issue better than I did before. Those scripts actually come from the RSC payload, so using RSCs is actually the case where the duplicate happens (which is a key piece I was missing earlier). But the idea remains the same, I wonder if you could use an SSR'd client component to get the schema data so there is no duplicate. I'm still working on figuring that out though. Using a client component would eliminate the self.__next payload for that component I'm pretty sure and as long as you pass the initial data needed for the schema to the client component as a prop so that it can be rendered server side, we should be able to eliminate this issue. The reason it shows up for server components is because the RSC payload includes the actual rendered output and data, but if it was a client component that was SSR'd, the payload would only need hydration information and references for the client component and thus should not have the duplicate schema data. 

I know I'm still speculating, I am just working on a project that I know will get to the point of having to figure this challenge out though. So I'm interested in solving it if possible before we get to it and would be interested in any solution you find.

**Reply by m4w** [reply]
Thanks, I will inspect it in detail now.

**Reply by m4w** [reply]
The solution was quite easy for me:
We build a simple Client-Component like this (as an example):
```javascript
'use client'

const StructuredDataWrapper = ({ data }) => {
  return <script type="application/ld+json">{JSON.stringify(data)}</script>
}

export default StructuredDataWrapper
```

This `data` object contains the complete structured data object, which can be set in a server component like this (another example):

```javascript
import StructuredDataWrapper from './StructuredDataWrapper'

export const StructuredOrganisationData = ({ siteConfig }) => {
  const organisation = {
    '@type': 'Organization',
    name: siteConfig.organisation_name,
    url: siteConfig?.organisation_url,
    address: {
      '@type': 'PostalAddress',
      streetAddress: siteConfig?.organisation_street_address,
      addressLocality: siteConfig?.organisation_address_locality,
      postalCode: siteConfig?.organisation_zip_code,
      addressCountry: siteConfig?.organisation_country,
    },
    contactPoint: [
      {
        '@type': 'ContactPoint',
        telephone: siteConfig?.organisation_telefone,
        contactType: siteConfig?.organisation_contact_type,
        email: siteConfig?.organisation_email,
      },
    ],
  }

  return <StructuredDataWrapper data={organisation} />
```

See [How are Server Components rendered?](https://nextjs.org/docs/app/building-your-application/rendering/server-components#how-are-server-components-rendered) for more information.

It is quite simple, if this part `<script type="application/ld+json">{JSON.stringify(data)}</script>` is rendered on the server, nextjs has to hydrate the client, including information about the script tag. This information can be picked up by the crawlers.

But if the script-Tag is inside of a client component, the hydration script contains only the props for `StructuredDataWrapper` and the crawlers are not able to pick that information. 

I hope that helps, although it is basically the same that @blf885 said.

**Reply by blf885** [reply]
That is exactly what I was thinking! Glad to hear that it works. I was reading some of the documentation while working on metadata and happened across this in the documentation: https://nextjs.org/docs/app/building-your-application/optimizing/metadata#json-ld 
It, unfortunately, does not mention anything about server/client components. It seems to me that this documentation should be updated to indicate that is should be used in a client component as you did. I'm just getting to the point in my project where I am implementing this, so I will report if I find anything else useful, but it seems that we have found a good solution without having to move to a new framework. I would like to see them improve the "generateMetadata" API to handle structured Data or at least provide a way to get it in the document's head. The Page Router had ways to have more control over the head, but it seems right now we are forced to have structured data in the body, which is fine according to Google's documentation, but not ideal for separation of concerns and later parsing.

**Reply by blf885** [reply]
After implementing some structured data, I'm not actually able to reproduce this issue at all. Interestingly, regardless of whether I use the solution we came up with using a client vs server component. Either way, I don't have the issue with the Rich Results test whether the HTML contains the  __next_f.push script with the data or not. Is it possible that the way Google parses the page for structured data has been updated? In order to get the __next_f.push script not to show up, I had to move the definition of the "const organisation" from your example inside the client wrapper, otherwise, the data would still obviously show up in the __next_f.push script. So in all honesty, I'm not sure why your solution works. But the fact that it does work combined with my tests using the Rich Result Test tool even when the __next_f.push script contains the data as well is what makes me wonder if Google made changes on their end. Additionally, I was never able to get schema.org's validator to pick up duplicates either, but never was OP.

**Comment by valley-boy** [comment]
If it any help.. we found no way round it an rebuilt using Astro. Sorry certainly not a quick fix but a solution!

**Reply by blf885** [reply]
Oh wow, was the schema data really that big of an issue for you guys? I may be misunderstanding its impact, but it doesn't seem duplicate schema data is that critical for SEO.

**Reply by muneeb-virtus** [reply]
yes it is really big for seo perspective, mostly when the competition is too high we really need this to win the race in search engines, i am also facing the same issue and came across this discussion, but sadly i end up with the Astro solution.

**Reply by ColeTownsend** [reply]
@valley-boy did you find that improved site ranking?

**Comment by itsmidhunnair** [comment]
Hi @valley-boy,

I ran into the same issue and wanted to add a bit more context.  
My Next.js application was working fine without the duplicated `schema.json`, until one day I noticed extra entries appearing in the schema validator tool.  

After some late-night debugging, I decided to break down the app and add components back one by one and finally found the cause. 

One of my component was created initially in server and later on was hydrated in the client side, and that component was using a common function for formatting date. The `formatDate` was using the `new Date()` function in it.  

Here’s what’s happening:
- The `new Date()` function behaves differently on the server and client.
- When it’s used in both server and client components, Next.js includes related data in the hydration chunk (`__NEXT_DATA__`).
- This results in the schema.json being duplicated (as formatDate was also being used in the function generating schema.json) the second instance being injected by the chunk.

Hope this helps someone facing the same issue!

Thanks!

**Reply by aryanjand** [reply]
Hi @itsmidhunnair,

I ran into the same issue everything worked fine, then suddenly I started seeing duplicated JSON-LD schema entries.

In my case, it wasn’t `new Date()`, but a change to how a toast portal was mounted in the app. I had switched from `document.body` → `window.top.document.body` (to support iframes).

It caused the duplicates because:
* Introduced a hydration-time state change (`useEffect` → re-render)
* Made the portal target non-deterministic across environments
* This destabilized hydration, causing Next.js to remount/replay parts of the tree

Even though this component doesn’t generate the JSON-LD schema entries, it triggered duplication on any pages using it.

* Server render → first schema
* Client replay → second schema

The main takeaway is the issue may not be in your schema code any hydration instability can cause this. Hope this helps!

**Comment by sefin23** [comment]
Synthesizing everything in this thread into a definitive answer, since there are now several confirmed solutions and root cause explanations:

## Root Cause

This is a **React Server Components (RSC) hydration behavior** issue, not a Next.js bug per se. When a Server Component renders a `<script type="application/ld+json">` tag, Next.js must include that rendered output in the RSC payload (`self.__next_f.push(...)`) so the client can hydrate/reconcile the tree. Crawlers that execute JavaScript (like Googlebot and the Rich Results Testing Tool) see **both** the original HTML tag and the hydration payload — and parse the JSON-LD from both, resulting in duplicates.

Additionally, as @itsmidhunnair and @aryanjand found, **any hydration instability** (e.g. `new Date()` behaving differently server vs. client, non-deterministic portals, etc.) can cause React to remount parts of the tree, producing a second render of your schema script.

## Solutions (in order of recommendation)

### 1. Wrap JSON-LD in a Client Component (Confirmed fix by @m4w)

Client Components only send hydration references — not the full rendered HTML — in the RSC payload. The `<script>` tag appears once in the initial HTML and the hydration chunk does not contain the schema data:

```tsx
// components/JsonLd.tsx
'use client'

interface JsonLdProps {
  data: Record<string, unknown>
}

export default function JsonLd({ data }: JsonLdProps) {
  return (
    <script
      type="application/ld+json"
      dangerouslySetInnerHTML={{ __html: JSON.stringify(data) }}
    />
  )
}
```

Then use it from a Server Component, passing the data as a prop:

```tsx
// app/product/[id]/page.tsx (Server Component)
import JsonLd from '@/components/JsonLd'

export default async function ProductPage({ params }) {
  const product = await getProduct(params.id)

  const schema = {
    '@context': 'https://schema.org',
    '@type': 'Product',
    name: product.name,
    // ...
  }

  return (
    <>
      <JsonLd data={schema} />
      {/* rest of page */}
    </>
  )
}
```

**Important:** Keep the schema data construction in the Server Component (as a prop). If you define it inside the Client Component, Next.js still needs to send it via the hydration payload.

### 2. Fix hydration mismatches in your app

If you start seeing duplicates after a code change that doesn't touch schema code, look for hydration instability:

- **`new Date()` / `Date.now()`** — produces different values on server vs. client. Use `suppressHydrationWarning` or compute the date server-side and pass as a string prop
- **`Math.random()`** — always different server vs. client
- **Browser-only APIs** in render (e.g., `window`, `document`, `localStorage`) without `typeof window !== 'undefined'` guards
- **Non-deterministic portal targets** (as @aryanjand found with `window.top.document.body`)

React's strict hydration rules mean any mismatch can cause a client replay that renders your schema script a second time.

### 3. Move JSON-LD to `<head>` via `generateMetadata`

If you want the schema in `<head>` (avoiding the body/hydration issue entirely), `generateMetadata` runs before streaming and its output lands in the document shell:

```ts
export async function generateMetadata({ params }): Promise<Metadata> {
  const product = await getProduct(params.id)
  return {
    other: {
      'script:ld+json': JSON.stringify({
        '@context': 'https://schema.org',
        '@type': 'Product',
        name: product.name,
      }),
    },
  }
}
```

Note: The `other` metadata field behavior for injecting script tags varies by Next.js version — test this in your specific version.

## Quick Diagnosis Checklist

1. Open DevTools → Network tab → view the page HTML source
2. Search for `__next_f.push` — if you see your schema data embedded in a string there, that’s the RSC payload causing the duplicate
3. Check for hydration warnings in the browser console (`Warning: Text content did not match`)
4. If you see hydration warnings, fix those first — they’re usually the real cause

## TL;DR

The fastest fix: move your `<script type="application/ld+json">` into a `'use client'` wrapper component and pass schema data as a prop from the Server Component. This prevents the schema from appearing in the RSC hydration payload.

**Comment by Brainstormex** [comment]
when i validate my website using the schema validator then i get the duplicate schema but when i paste the view page source code then the duplicate schema doesnt appear why
when i search the dom then i see the duplicated schema in the DOM also, not able to figure this out

**Reply by abhi2c** [reply]
Similar issue here, when i inspect the page i see my normal <script jsonld> and then the self.f.push rsc payload.. so when i run it through rich results its only flagging out aggregateRating key.. not sure what the issue is
